File: Coding/Blend/数据结构/二叉树/AVLTree.cpp

#include<iostream>
#include <cmath>
#include <algorithm>
using namespace std;

class AVLTree 
{
public:
	int val;
	AVLTree* left;
	AVLTree* right;

	AVLTree(int val)
	{
		this->val = val;
		this->left = NULL;
		this->right = NULL;
	}

	 AVLTree* insertTreeNode(AVLTree* node, int val)
	{
		if (node == NULL)
		{
			node = new AVLTree(val);
		}

		if (val < node->val)
		{
			// 插入左子树 

			node->left = insertTreeNode(node->left, val);
			if (getTreeHeight(node->left) - getTreeHeight(node->right) == 2)
			{
				if (val < node->left->val)
				{
					// 左左状态   右单旋 
					node = singleRightRotate(node);
				}
				else if (val > node->left->val)
				{
					// 左右状态  左单旋 后右单旋 
					node = leftRightRotate(node);
				}
			}
		}
		else if (val > node->val)
		{
			// 插入右子树 

			node->right = insertTreeNode(node->right, val);

			if (getTreeHeight(node->right) - getTreeHeight(node->left) == 2)
			{
				if (val < node->right->val)
				{
					// 右左状态 先右单旋 后左单旋 
					node = rightLeftRotate(node);
				}
				else if (val > node->right->val)
				{
					// 右右状态 左单旋 
					node = singleLeftRotate(node);
				}
			}
		}
		return node;
	}

private:
	AVLTree* singleLeftRotate(AVLTree* node)
	{
		AVLTree* right = node->right;
		node->right = right->left;
		right->left = node;
		return right;
	}

	AVLTree* singleRightRotate(AVLTree* node)
	{
		AVLTree* left = node->left;

		node->left = left->right;

		left->right = node;

		return left;
	}

	AVLTree* leftRightRotate(AVLTree* node)
	{
		node->left = singleLeftRotate(node->left);
		node = singleRightRotate(node);
		return node;
	}

	AVLTree* rightLeftRotate(AVLTree* node)
	{
		node->right = singleRightRotate(node->right);
		node = singleLeftRotate(node);
		return node;
	}

	// 获取tree的高度 
	int getTreeHeight(AVLTree* tree)
	{
		if (tree == NULL)
		{
			return 0;
		}

		return max(getTreeHeight(tree->left), getTreeHeight(tree->right)) + 1;
	}
};

void InOrderTranverse(AVLTree* root)
{
	if (root == NULL)
	{
		return;
	}

	InOrderTranverse(root->left);

	cout << root->val << endl;

	InOrderTranverse(root->right);
}

void PreTranverse(AVLTree* root)
{
	if (root == NULL)
	{
		return;
	}

	cout << root->val << endl;
	PreTranverse(root->left);
	PreTranverse(root->right);
}

void AfterTranverse(AVLTree* root)
{
	if (root == NULL)
	{
		return;
	}

	AfterTranverse(root->left);
	AfterTranverse(root->right);
	cout << root->val << endl;
}

int main()
{
	AVLTree* root = new AVLTree(1);
	root = root->insertTreeNode(root, 2);
	root = root->insertTreeNode(root, 3);
	root = root->insertTreeNode(root, 4);
	root = root->insertTreeNode(root, 5);
	root = root->insertTreeNode(root, 6);
	root = root->insertTreeNode(root, 7);
	root = root->insertTreeNode(root, 8);
	root = root->insertTreeNode(root, 9);
	root = root->insertTreeNode(root, 10);
	root = root->insertTreeNode(root, 11);
	root = root->insertTreeNode(root, 12);
	root = root->insertTreeNode(root, 13);
	root = root->insertTreeNode(root, 14);
	root = root->insertTreeNode(root, 15);

	cout << "Pre tranverse" << endl;
	PreTranverse(root);

	cout << "InOrder tranverse" << endl;
	InOrderTranverse(root);

	cout << "After tranverse" << endl;
	AfterTranverse(root);
	return 0;
}

File: Coding/computer/数据结构/二叉树/AVLTree.py

def getHeight(tree):
    if tree == None:
        return 0

    return max(getHeight(tree.left), getHeight(tree.right)) + 1

class AVLTree(object):
    def __init__(self, val, left, right):
        self.val = val
        self.left = left
        self.right = right

    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

    def rightSingleRotate(self, root):
        temp = root.left
        root.left = temp.right
        temp.right = root

        return temp

    def leftSingleRotate(self, root):

        temp = root.right
        root.right = temp.left
        temp.left = root

        return temp

    def leftRightRotate(self, root):
        root.left = self.leftSingleRotate(root.left)
        root = self.rightSingleRotate(root)
        return root

    def rightLeftRotate(self, root):
        root.right = self.rightSingleRotate(root.right)
        root = self.leftSingleRotate(root)
        return root

    def insertNode(self, root, val):
        if root == None:
            root = AVLTree(val)

        if val < root.val:
            # 插入左子树
            root.left = self.insertNode(root.left, val)

            if getHeight(root.left) - getHeight(root.right) == 2:
                # 调整树的结构
                if val < root.left.val:
                    # 右单旋
                    root = self.rightSingleRotate(root)

                elif val > root.left.val:
                    # 左右旋转
                    root = self.leftRightRotate(root)

            pass
        elif val > root.val:
            # 插入右子树
            root.right = self.insertNode(root.right, val)

            if getHeight(root.right) - getHeight(root.left) == 2:
                if val < root.right.val:
                    # 右左旋转
                    root = self.rightLeftRotate(root)
                elif val > root.right.val:
                    # 左单旋
                    root = self.leftSingleRotate(root)

        return root

    def inorderTranverse(self, root):
        if root == None:
            return

        self.inorderTranverse(root.left)
        print(root.val)
        self.inorderTranverse(root.right)



def main():
    root = AVLTree(1)
    root = root.insertNode(root, 2)
    root = root.insertNode(root, 3)
    root = root.insertNode(root, 4)
    root = root.insertNode(root, 5)
    root = root.insertNode(root, 6)
    root = root.insertNode(root, 7)
    root = root.insertNode(root, 8)
    root = root.insertNode(root, 9)
    root = root.insertNode(root, 10)
    root = root.insertNode(root, 11)
    root = root.insertNode(root, 12)
    root = root.insertNode(root, 13)
    root = root.insertNode(root, 14)
    root = root.insertNode(root, 15)

    root.inorderTranverse(root)


if __name__ == "__main__":
    main()
